app/javascript/gameStatesFactory.js

```diff
@@ -0,0 +1,304 @@
+    delete gameStates.selfMoves[gameId];
+    delete gameStates.opponentMoves[gameId];
+    delete gameStates.lastMoveNumber[gameId];
+
+    gameStates.updateLocalStorage();
+  };
+
+  gameStates.getSelfMoves = function(gameId) {
+    if(typeof(gameStates.selfMoves[gameId]) === 'undefined') {
+      return [];
+    }
+
+    return gameStates.selfMoves[gameId];
+  };
+  gameStates.getLastSelfMove = function(gameId) {
+    if(typeof(gameStates.selfMoves[gameId]) === 'undefined') {
+      return [];
+    }
+    if(gameStates.selfMoves[gameId].length === 0) {
+      throw 'This game has no self moves yet';
+    }
+
+    return gameStates.selfMoves[gameId][gameStates.selfMoves[gameId].length - 1];
+  };
+  gameStates.getSelfMove = function(gameId, moveNumber) {
+    if(typeof(gameStates.selfMoves[gameId]) === 'undefined') {
+      throw 'This game has no moves yet';
+    }
+    if(gameStates.selfMoves[gameId].length === 0) {
+      throw 'This game has no self moves yet';
+    }
+
+    for(let currentMove of gameStates.selfMoves[gameId]) {
+      let currentMoveNumber = gameStates.getMoveNumberFromState(currentMove.newState);
+
+      if(currentMoveNumber === moveNumber) {
+        return currentMove;
+      }
+    }
+  };
+
+  gameStates.getPreviousOpponentMove = function(gameId) {
+    if(typeof(gameStates.opponentMoves[gameId]) === 'undefined') {
+      throw 'This game has no moves yet';
+    }
+    if(gameStates.opponentMoves[gameId].length === 0) {
+      throw 'This game has no opponent moves yet';
+    }
+    if(gameStates.opponentMoves[gameId].length < 2) {
+      throw 'This game has no previous opponent move';
+    }
+  };
+  gameStates.getLastOpponentMove = function(gameId) {
+    if(typeof(gameStates.opponentMoves[gameId]) === 'undefined') {
+      throw 'This game has no moves yet';
+    }
+    if(gameStates.opponentMoves[gameId].length === 0) {
+      throw 'This game has no opponent moves yet';
+    }
+
+    return gameStates.opponentMoves[gameId][gameStates.opponentMoves[gameId].length - 1];
+  };
+
+  /**
+   *
+   * @param gameId
+   * @return boolean
+     */
+  gameStates.isBlockchainStateNewer = function(gameId) {
+    let blockchainState = Chess.getCurrentGameState(gameId);
+
+    if(blockchainState) {
+      if(typeof(gameStates.lastMoveNumber[gameId]) === 'undefined') {
+        // No state in the local states, so return true
+        return true;
+      }
+
+      let blockchainMoveNumber = gameStates.getMoveNumberFromState(blockchainState);
+
+      return blockchainMoveNumber > gameStates.lastMoveNumber[gameId];
+    }
+    // No state in the blockchain, so return false
+    return false;
+  };
+
+  /**
+   * Get an array with all parameters needed to send them to the blockchain.
+   * Throws if the last move was not done by oneself.
+   * @param gameId
+   * @return [oldOpponentState, oldOpponentStateSignature, moveSelfFrom, moveSelfTo]
+     */
+  gameStates.getSelfMovePackage = function(gameId) {
```

Comment:
> Throws if the last move was not done by oneself.
We do not want to throw if the last move was done by the opponent. If this is the case, we want the package to be our last move with the state the move was based on.

181 should not throw ;)